lib/request_overrider.js

```diff
@@ -289,92 +303,115 @@ class InterceptedRequestHandler {
             : originalHttpRequest(options)
 
         propagate(newReq, req)
-        //  We send the raw buffer as we received it, not as we interpreted it.
+        // We send the raw buffer as we received it, not as we interpreted it.
         newReq.end(requestBodyBuffer)
-        return
+      } else {
+        const err = new Error(
+          `Nock: No match for request ${common.stringifyRequest(
+            options,
+            requestBodyString
+          )}`
+        )
+        err.statusCode = err.status = 404
+        this.emitError(err)
       }
-
-      const err = new Error(
-        `Nock: No match for request ${common.stringifyRequest(
-          options,
-          requestBodyString
-        )}`
-      )
-      err.statusCode = err.status = 404
-      this.emitError(err)
-      return
     }
+  }
+}
 
-    debug('interceptor identified, starting mocking')
+/**
+ * Play back an intercepto using the given request and mock response.
+ */
+class InterceptorPlayback {
+  constructor({
+    req,
+    socket,
+    options,
+    requestBodyString,
+    isBinaryRequestBodyBuffer,
+    response,
+    interceptor,
+  }) {
+    this.req = req
+    this.socket = socket
+    this.options = options
+    this.requestBodyString = requestBodyString
+    this.response = response
+    this.interceptor = interceptor
+    this.isBinaryRequestBodyBuffer = isBinaryRequestBodyBuffer
+  }
```

Comment:
I don't see why this should be a class. It could just be a single fn that accepts the args currently coming into the contractor and holds the logic of `start`. The methods defined on this class could then just be nested functions. That would also clean up all the `this` deconstructing. 

The bottom half of this file could also be moved to its own file, but I would ask that that happen in a different PR is folks like the idea.

Reply:
> I don't see why this should be a class. It could just be a single fn that accepts the args currently coming into the contractor and holds the logic of `start`. The methods defined on this class could then just be nested functions. That would also clean up all the `this` deconstructing.

Good point.

I use this pattern a lot: very simple classes with constructors which assign properties and methods which access them. I think it has some advantages over a closure. The destructuring makes the state dependencies of each little method really obvious, whereas a closure requires a bit more studying. It also makes clearer the difference between calls to another method that might rely on other state, vs a function imported or defined elsewhere, whereas with a closure optically these are the same.

In this case though, I don't think anything is gained, especially as there's no real instance state. Agree that converting this to a closure will improve it.


> The bottom half of this file could also be moved to its own file, but I would ask that that happen in a different PR is folks like the idea.

Totally agreed.

Reply:
I've converted the InterceptorPlayback class to a playbackInterceptor function. I left the `start()` function intact for the moment – it's invoked at the bottom of the function. Both `playbackInterceptor()` and `start()` are pretty long, and I think this is all still very difficult to follow.

I'm inclined to continue to break this down into smaller logical units, and also to rewrite it using `async`. Though I think the next step after this should be to move it into its own file before continuing with further refactors.